src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1386,6 +1413,37 @@ public IList<string> ValidValues
         {
             get
             {
+                if (validValuesGeneratorType != null)
+                {
+                    ValidValuesCacheEntry validValuesCacheEntry;
+                    if (validValuesCache.TryGetValue(validValuesGeneratorType, out validValuesCacheEntry))
+                    {
+                        var currentTime = DateTime.Now;
+                        if (DateTime.Compare(validValuesCacheEntry.cacheEntryCreationTime.AddSeconds(CacheExpiration), currentTime) < 0)
+                        {
+                            // Update expired Entry.
+                            validValuesCacheEntry.validValues = validValuesCacheEntry.validValuesGenerator.GetValidValues()?.ToArray();
+                            validValuesCacheEntry.cacheEntryCreationTime = currentTime;
+                            validValuesCache[validValuesGeneratorType] = validValuesCacheEntry;
+                        }
+                    }
+                    else
+                    {
+                        var validValuesGenerator = (IValidateSetValuesGenerator)Activator.CreateInstance(validValuesGeneratorType);
+                        var validValues = validValuesGenerator.GetValidValues()?.ToArray();
+                        var cacheEntryCreationTime = DateTime.Now;
+                        validValuesCacheEntry = new ValidValuesCacheEntry(validValuesGenerator, validValues, cacheEntryCreationTime);
+                        validValuesCache.TryAdd(validValuesGeneratorType, validValuesCacheEntry);
+                    }
+                    if (validValuesCacheEntry.validValues == null)
+                    {
+                        throw new ValidationMetadataException(
+                            "ValidateSetGeneratedValidValuesListIsEmpty",
+                            null,
+                            Metadata.ValidateSetGeneratedValidValuesListIsEmpty);
+                    }
+                    _validValues = validValuesCacheEntry.validValues;
```

Comment:
>Is there a good reason to store the values in the field instead of  return validValuesCacheEntry.validValues; ?

We only use this in `SetAsString()` for error message. We could use `ValidValues` if we're not afraid that there will be a cache expiration and a recalculation of  `ValidValues` and that there may be other data in the error message.

>  Task.Delay(CacheExpiration)

Different attributes can have different `CacheExpiration`. How can we resolve the problem?